lsp-completion.el

```diff
@@ -0,0 +1,542 @@
+                  additional-text-edits?)
+              (lsp--apply-text-edits additional-text-edits?)
+            (-let [(callback cleanup-fn) (lsp--create-apply-text-edits-handlers)]
+              (lsp--resolve-completion-async
+               item
+               (lambda (resolved-item)
+                 (funcall callback
+                          (lsp:completion-item-additional-text-edits? resolved-item)))
+               cleanup-fn))))
+
+        (when (and lsp-signature-auto-activate
+                   (lsp-feature? "textDocument/signatureHelp"))
+          (lsp-signature-activate))
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+;;;###autoload
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+  (when (and lsp-completion-enable
+             (lsp-feature? "textDocument/completion"))
+    (lsp-completion-mode 1)))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (setq-local completion-at-point-functions nil)
+    (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+    (setq-local completion-category-defaults
+                (add-to-list 'completion-category-defaults '(lsp-capf (styles basic))))
+
+    (when (lsp--capability :completionProvider)
+      (cond
+       ((and (functionp 'company-lsp)
+             (not lsp-completion-prefer-capf))
+        (progn
+          (company-mode 1)
+          (add-to-list 'company-backends 'company-lsp)
+          (setq-local company-backends (remove 'company-capf company-backends))))
+
+       ((and (fboundp 'company-mode) lsp-completion-enable)
+        (company-mode 1)
+        (add-to-list 'company-backends 'company-capf))))
+
+    (when (bound-and-true-p company-mode)
+      (add-hook 'company-completion-started-hook
+                (lambda (&rest _)
+                  (setq-local lsp-inhibit-lsp-hooks t))
+                nil
+                t)
+      (add-hook 'company-after-completion-hook
+                (lambda (&rest _)
+                  (lsp-completion--capf-clear-cache)
+                  (setq-local lsp-inhibit-lsp-hooks nil))
+                nil
+                t))
+    (add-hook 'lsp-configure-hook #'lsp-completion--enable nil t)
+    (add-hook 'lsp-unconfigure-hook #'lsp-completion--disable nil t))
+   (t
+    (remove-hook 'completion-at-point-functions #'lsp-completion-at-point t)
+    (setq-local completion-category-defaults
+                (cl-remove 'lsp-capf completion-category-defaults :key #'car))
+    (remove-hook 'lsp-configure-hook #'lsp-completion--enable t)
```

Comment:
the same - if we are adding it via autoload

Reply:
I think if we remove the `remove-hook`, the user won't be able to disable this even turning off the mode, right?

Reply:
This operation atm is NOP, because this is not removing the hook that is added in the end of the file.

Reply:
Ok, it makes sense, done!